content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md

```diff
@@ -0,0 +1,76 @@
+---
+layout: blog
+title: "Eviction policy for unhealthy pods guarded by PodDisruptionBudgets"
+date: 2023-01-06
+slug: "unhealthy-pod-eviction-policy-for-pdbs"
+---
+
+**Authors:** Filip Křepinský (Red Hat), Morten Torkildsen (Google), Ravi Gudimetla (Apple)
+
+## What problems does this solve?
+
+Eviction of pods respects Pod Disruption Budgets (PDBs). This means that the eviction of a pod
+should not disrupt a guarded application and `.status.currentHealthy` of a PDB should not fall 
+bellow `.status.desiredHealthy`. Running pods that are [Unhealthy](/docs/tasks/run-application/configure-pdb/#healthiness-of-a-pod)
+do not count towards the PDB status, but eviction of these is only possible in case the application
+is not disrupted. This helps disrupted or not yet started application to achieve availability
+as soon as possible without additional downtime that would be caused by evictions.
+
+Unfortunately, this poses a problem for cluster administrators that would like to drain nodes
+without any manual interventions. Misbehaving applications with pods in `CrashLoopBackOff`
+state (due to a bug or misconfiguration) or pods that are simply failing to become ready
+make this task much harder.
+
+On the other hand there are users that depend on the current behavior to:
+- prevent data-loss that would be caused by deleting pods that are guarding an underlying resource or storage
+- achieve the best availability possible for their application
+
+We are introducing a new field `spec.unhealthyPodEvictionPolicy` for PDBs to support both of these requirements.
+
+## How does it work?
+
+There are two policies `IfHealthyBudget` and `AlwaysAllow` to choose from. 
+The former conforms to the original behavior to achieve the best availability.
+This will be also used when no policy is specified.
+
+By setting `AlwaysAllow` policy to the `spec.unhealthyPodEvictionPolicy` field of your PDB, 
+you are choosing the best effort availability for your application.
+This will make it easier to maintain and upgrade your clusters.
+
+Eviction API will consider these policies when eviction of a pod that is guarded by a PDB is requested. 
```

Comment:
I am not sure if I follow.

we are describing the behaviour of the default IfHealthyBudget in the section above with.

```
Eviction of pods respects PodDisruptionBudgets (PDBs). This means that the eviction of a pod
should not disrupt a guarded application and `.status.currentHealthy` of a PDB should not fall 
below `.status.desiredHealthy`. Running pods that are [Unhealthy](/docs/tasks/run-application/configure-pdb/#healthiness-of-a-pod)
do not count towards the PDB status, but eviction of these is only possible in case the application
is not disrupted. This helps disrupted or not yet started application to achieve availability
as soon as possible without additional downtime that would be caused by evictions.
```

or are you suggesting we should describe the edge cases of Eviction API?

Reply:
If:
- you have a workload with _n_ Pods
- you protect the workload with a PodDisruptionBudget
- you choose, or default to, `IfHealthyBudget`
- several of those Pods are in CrashLoopBackoff

and you then
- (attempt to) drain a node where the CrashLoopBackoffs are happening

what actually happens? It'd be good to summarise this mainly in terms of `kubectl drain`, but readers might also want to know at API level. For example: you try to create an eviction, and that request fails with error 418, `I'm a teapot` (OK, probably not that specific error code).

My ask here is to explain the outcome that you _don't_ see if you enable the feature gate and set a custom `unhealthyPodEvictionPolicy`, so that readers can appreciate the benefit.

Reply:
I see, I have tried to put more details to describe the behavior.

```
Unfortunately, this poses a problem for cluster administrators that would like to drain nodes
without any manual interventions. Misbehaving applications with pods in `CrashLoopBackOff`
state (due to a bug or misconfiguration) or pods that are simply failing to become ready
make this task much harder. Any eviction request will fail due to violation of a PDB, 
when all pods of an application are unhealthy. Draining of a node cannot make any progress
in that case.
```

Do you think it make sense to describe the exact message and http code here?

Reply:
At this point, you should describe _something_. I haven't tried the feature out so I don't know what you need to say. I don't think “Eviction API will consider these policies when eviction of a pod that is guarded by a PDB is requested.” feels like enough.

That said, if we had to we could merge this and live with the confusion.

Reply:
Thx for the suggestion, I used part of it and reorganized the section: https://github.com/kubernetes/website/compare/4629fb1dcda1ab70deaed7a30f89228d0d1fbb6d..e27c6086386bcbdc67b52cf7c9909d3c5caf788d.

I think we should not mention the node-pressure eviction here, since it could distract from the main topic. The node-pressure eviction does not abide by the PDBs. I have tried to highlight that we are dealing with the API initiated eviction.